### PR TITLE
Added a callback mechanism when creating / updating a form via the builder

### PR DESCRIPTION
**Pull request by brunobord**
TODO 

- [x] factorize redundant code
- [x] don't use the ``func`` variable in logs, rather the namespace string.

### CODE REVIEW

File: formidable/views.py [hunk 1]
```diff
@@ -71,6 +77,40 @@ class FormidableCreate(six.with_metaclass(MetaClassView, CreateAPIView)):
...
 
+    def dispatch(self, request, *args, **kwargs):
+        # Being forced to do this in dispatch() rather than post() because
+        # ValidationErrors can be raised in dispatch and we don't go through
+        # the post() method
+        response = super(FormidableCreate, self).dispatch(
+            request, *args, **kwargs)
+
+        # Is the response a success or a failure?
+        if request.method == 'POST':
```
Comment (moumoutte) [inline formidable/views.py]: instead of using it at the dispatch time, I suggest you to override the perform_create and perform_update method on the success moment.

Like you can see on rest_framework source code,  Create and Update views have both this kind of method.

https://github.com/tomchristie/django-rest-framework/blob/master/rest_framework/mixins.py#L25
https://github.com/tomchristie/django-rest-framework/blob/master/rest_framework/mixins.py#L79

In the source code you can easily see when your are in this method, you know the creation or update is a success.

For the create example, I suggest a code like this.

```python

    def perform_create(self, serializer):
        super(MyAwesomeCreateView, self).perform_create()
        self.call_formidable_callback() # The method name is up to you :p

```

For the failure moment, I suggest to use the "handle_exception" contains in rest framework view

https://github.com/tomchristie/django-rest-framework/blob/master/rest_framework/views.py#L422


```python

   def handle_exceptions(self,  exc):
        response = super(MyAwesomeCreateView, self).handle_exceptions(exc)
        self.call_formidable_errback()
        return response

```

We can imagine a Dynamic Mixin View to handle properly loading those (call/err)back


For information, don't hesitate to use the rest_framework helpers to load module, it's powerful !



```python

from rest_framework.settings import import_from_string

class MyMixinFormidableCallBack(object:)

       callback_settings_name = None # to implement at the final class level
       errback_settings_name = None

       def __init__(self, *args, **kwargs):
   
                self.callback_meth = import_from_string(settings.get(callback_settings_name), self.callback_settings_name)

                self.errback_meth = import_from_string(settings.get(callback_settings_name), self.callback_settings_name)

      def call_formidable_callback(self):
             if self.callback_meth:
                  return self.callback_meth()
              return None

       
      def call_formidable_errback(self):
             if self.errback_meth:
                  return self.callback_meth()
              return None
            
```

Sorry for the long comment, this is "code out of my head" I didn't test it, it's just a idea about implementation. 

But the way you built it seems correct to me, so many thanks for this POC and complete test.

**Comment by brunobord** [comment]
ready for review

File: formidable/views.py [hunk 2]
```diff
@@ -21,6 +24,86 @@
...
+        """
+        Tool to simply call the callback function and handle edge-cases.
+        """
+        func = extract_function(callback)
+        # Call function only if existing
+        if not func:
+            return
+        try:
+            func(self.request)
+        except Exception:
```
Comment (moumoutte) [inline formidable/views.py]: I'm not really convince by catching all exception on do not log the traceback. I'm  more confortable with let the exception escape.

File: formidable/views.py [hunk 3]
```diff
@@ -21,6 +24,86 @@
 from formidable.serializers.presets import PresetsSerializer
 
 
+logger = logging.getLogger(__name__)
+
+
+def extract_function(func_name):
```
Comment (moumoutte) [inline formidable/views.py]: indead of using this, the best thing to do is to use the import module writing by django rest framework.

File: formidable/views.py [hunk 4]
```diff
@@ -21,6 +24,86 @@
...
+        # Extract the callback function
+        callback = getattr(settings, self.failure_callback_settings, None)
+        self._call_callback(callback)
+
+    def perform_create(self, serializer):
+        response = super(CallbackMixin, self).perform_create(serializer)
+        self.success_callback()
+        return response
+
+    def perform_update(self, serializer):
```
Comment (moumoutte) [inline formidable/views.py]: I would like to restore the method in the appropriate class. No need `perform_update` on CreateView.

File: formidable/views.py (hunk 3 above)
Comment (brunobord) [inline formidable/views.py]: probably, yes. I'll try it out.

File: formidable/views.py (hunk 2 above)
Comment (brunobord) [inline formidable/views.py]: the problem is that if you misconfigure your settings value, you'll only get this exception at runtime, and not before. **Every** form updated will crash, and every update view will throw a 500 - the db transaction will probably be rolled back so your data won't be recorded ; and you'll need a redeployment with a modified settings value to fix this (which could mean a new release, etc).
I'd be sad to see that, even if my form data is valid, that the  ``form.save()`` has been rejected because of an operation that has nothing to do with me, but comes from a "post-update" operation. Let's imagine this as a celery task, *"fire and forget"*...

May I suggest that, in the next iteration of this feature, we could eventually add a boolean settings: ``FORMIDABLE_IGNORE_EXCEPTIONS_ON_CALLBACK``, set to False by default. If set to True, the caught exception would be re-raised.

File: formidable/views.py (hunk 4 above)
Comment (brunobord) [inline formidable/views.py]: Do you mean that we'd have a ``CallbackMixin``, and two children: ``CreateCallbackMixin``, with ``post_create``, and ``UpdateCallbackMixin`` with ``post_update``?

File: formidable/views.py (hunk 3 above)
Comment (moumoutte) [inline formidable/views.py]: The good method can be : https://github.com/tomchristie/django-rest-framework/blob/master/rest_framework/settings.py#L169

File: formidable/views.py (hunk 2 above)
Comment (moumoutte) [inline formidable/views.py]: > The problem is that if you misconfigure your settings value, you'll only get this exception at runtime, and not before

The right response is to not allowed the loading of the module if the configuration is not okay with that.

By using the import_from_string module, the function through an exception with the reason of the fail (Bad configuration). We have to load at the module init to avoid exception at the runtime.

File: formidable/views.py (hunk 4 above)
Comment (moumoutte) [inline formidable/views.py]: No, I mean to right directly those method in the final view class.

File: formidable/views.py (hunk 2 above)
Comment (brunobord) [inline formidable/views.py]: I think I made a mistake... the exception we're talking about is not at import time, but at runtime. There are two separate possible errors:

* the module/function is missing, it's an import error. I agree, it can be handled at load time, loading functions at the top level of django-formidable and throwing an ImproperlyConfigured exception there.
* the function exists, is loaded properly, but for some reason, crashes / throws an error/exception : there, IMHO, you should not reject the form by re-raising the exception, because it's not the users fault, but the integrator's fault.

Comment (brunobord) [inline formidable/views.py]: the problem with throwing the exception at load time is that it's hard to test... :/

Comment (brunobord) [inline formidable/views.py]: note: the name of the settings variable is set in the class: https://github.com/novafloss/django-formidable/pull/137/files#diff-838bc4b57b8469118af87d088289ef93R139

As a consequence, it's overridable ; it would require a lot of spaghetti code, though, because this view is directly linked to the builder API urls... but it's doable. I mean we could check for a variable at load time, that won't be used at runtime.

We'll have to keep this in mind.

**Comment by brunobord** [comment]
ready for review again

**Comment by brunobord** [comment]
Added documentation + finally made it possible to use ``django.contrib.messages`` in callbacks as an example (and in the docs). Life is fine, fine, fine.

**Comment by brunobord** [comment]
post-save-message

**Comment by brunobord** [comment]
I'm not sure I'd want to squash all these commits into one ; I've let the important milestones in this history (e.g. adding a failing test and implementing the solution right after)
